subsys/bluetooth/services/ots/ots.c

```diff
@@ -271,63 +288,90 @@ static ssize_t ots_obj_prop_read(struct bt_conn *conn,
 				 sizeof(ots->cur_obj->metadata.props));
 }
 
-int bt_ots_obj_add(struct bt_ots *ots,
-			    struct bt_ots_obj_metadata *obj_init)
+int bt_ots_obj_add_internal(struct bt_ots *ots, struct bt_conn *conn,
+			    const struct bt_ots_obj_add_param *param,
+			    struct bt_gatt_ots_object **obj)
 {
 	int err;
-	struct bt_gatt_ots_object *obj;
-	size_t name_len;
+	struct bt_gatt_ots_object *new_obj;
+	struct bt_ots_obj_created_param created_param;
 
 	if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ) && ots->dir_list &&
 	    ots->dir_list->dir_list_obj->state.type != BT_GATT_OTS_OBJECT_IDLE_STATE) {
 		LOG_DBG("Directory Listing Object is being read");
 		return -EBUSY;
 	}
 
-	name_len = strlen(obj_init->name);
-
-	CHECKIF(name_len == 0 || name_len > CONFIG_BT_OTS_OBJ_MAX_NAME_LEN) {
-		LOG_DBG("Invalid name length %zu", name_len);
-		return -EINVAL;
-	}
-
-	CHECKIF(!ots_obj_validate_prop_against_oacp(obj_init->props, ots->features.oacp)) {
-		LOG_DBG("Object properties (0x%04X) are not a subset of OACP (0x%04X)",
-				obj_init->props, ots->features.oacp);
-		return -ENOTSUP;
-	}
-
-	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &obj);
+	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &new_obj);
 	if (err) {
 		LOG_ERR("No space available in the object manager");
 		return err;
 	}
 
-	/* Initialize object. */
-	memcpy(&obj->metadata, obj_init, sizeof(obj->metadata));
-
 	if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
-		bt_ots_dir_list_obj_add(ots->dir_list, ots->obj_manager, ots->cur_obj, obj);
+		bt_ots_dir_list_obj_add(ots->dir_list, ots->obj_manager, ots->cur_obj, new_obj);
 	}
 
-	/* Request object data. */
+	(void)memset(&created_param, 0, sizeof(created_param));
+
 	if (ots->cb->obj_created) {
-		err = ots->cb->obj_created(ots, NULL, obj->id, obj_init);
+		err = ots->cb->obj_created(ots, NULL, new_obj->id, param, &created_param);
+
 		if (err) {
-			bt_gatt_ots_obj_manager_obj_delete(obj);
+			bt_gatt_ots_obj_manager_obj_delete(new_obj);
 
 			if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
 				bt_ots_dir_list_obj_remove(ots->dir_list, ots->obj_manager,
-							   ots->cur_obj, obj);
+							   ots->cur_obj, new_obj);
 			}
 
 			return err;
 		}
 	}
 
+	new_obj->metadata.type = param->type;
+
+	new_obj->metadata.name = created_param.name;
+	new_obj->metadata.size = created_param.size;
+	new_obj->metadata.props = created_param.props;
+
+	__ASSERT(ots_obj_validate_prop_against_oacp(new_obj->metadata.props, ots->features.oacp),
+		 "Object properties (0x%04X) are not a subset of OACP (0x%04X)",
+		 new_obj->metadata.props, ots->features.oacp);
+
+	__ASSERT(new_obj->metadata.name != NULL,
```

Comment:
This ASSERT and the one above check if the created callback returned usable values.
Proposal: Change these from asserts to error checks, and move them alongside the check for whether the created callback returned an error.
If the callback returned an error, or if any of these two checks fail, then delete the object and return error.